In [8]:
import numpy as np

print("NumPy Data Explorer")

NumPy Data Explorer


In [9]:
data = np.genfromtxt(
    "Star_Wars_Characters.csv",
    delimiter=";",
    dtype=str,
    encoding="cp1252",
    skip_header=1
)

print(data)

[['Anakin Skywalker' 'Tatooine' '22' 'Republic ? Empire' 'Jedi / Sith']
 ['Padmé Amidala' 'Naboo' '27' 'Republic' 'Neutral (politician)']
 ['Obi-Wan Kenobi' 'Stewjon' '~38' 'Republic' 'Jedi']
 ['Qui-Gon Jinn' 'Unknown' 'Deceased' 'Republic' 'Jedi']
 ['Yoda' 'Unknown' '~880' 'Republic' 'Jedi']
 ['Mace Windu' 'Haruun Kal' '~53' 'Republic' 'Jedi']
 ['Palpatine / Darth Sidious' 'Naboo' '~65' 'Republic ? Sith' 'Sith']
 ['Count Dooku / Darth Tyranus' 'Serenno' '~83' 'Confederacy' 'Sith']
 ['Countess Dooku (cameo)' 'Serenno' '~80+' 'Confederacy' 'Neutral']
 ['Bail Organa' 'Alderaan' '~45' 'Republic' 'Neutral']
 ['Jar Jar Binks' 'Naboo' '~30' 'Republic' 'Neutral']
 ['Boss Nass' 'Naboo' '~60' 'Republic' 'Neutral']
 ['Shmi Skywalker' 'Tatooine' 'Deceased' 'Civilian' 'Neutral']
 ['Watto' 'Tatooine' '~50' 'Neutral' 'Neutral']
 ['Sebulba' 'Malastare' '~30' 'Neutral' 'Neutral']
 ['Captain Panaka' 'Naboo' '~40' 'Republic' 'Neutral']
 ['Captain Typho' 'Naboo' '~30–40' 'Republic' 'Neutral']
 ['Sabé' 'N

In [10]:
print("Number of rows:", data.shape[0])
print("Number of columns:", data.shape[1])
print("Array shape:", data.shape)
print("Array size:", data.size)

Number of rows: 94
Number of columns: 5
Array shape: (94, 5)
Array size: 470


In [11]:
# Indexing

print("First character:")
print(data[0])

print("\nSecond character:")
print(data[1])

print("\nLast character:")
print(data[-1])

First character:
['Anakin Skywalker' 'Tatooine' '22' 'Republic ? Empire' 'Jedi / Sith']

Second character:
['Padmé Amidala' 'Naboo' '27' 'Republic' 'Neutral (politician)']

Last character:
['Senator Tundra Dowmeia' 'Mon Calamari' '~45–50' 'Republic' 'Politician']


In [12]:
# Access individual values

print("First character name:", data[0, 0])
print("First character birthplace:", data[0, 1])
print("First character age:", data[0, 2])
print("First character affiliation:", data[0, 3])
print("First character status:", data[0, 4])

First character name: Anakin Skywalker
First character birthplace: Tatooine
First character age: 22
First character affiliation: Republic ? Empire
First character status: Jedi / Sith


In [13]:
# Slicing

print("First 5 rows:")
print(data[:5])

print("\nFirst 5 character names:")
print(data[:5, 0])

print("\nFirst 10 character names and ages:")
print(data[:10, [0, 2]])

First 5 rows:
[['Anakin Skywalker' 'Tatooine' '22' 'Republic ? Empire' 'Jedi / Sith']
 ['Padmé Amidala' 'Naboo' '27' 'Republic' 'Neutral (politician)']
 ['Obi-Wan Kenobi' 'Stewjon' '~38' 'Republic' 'Jedi']
 ['Qui-Gon Jinn' 'Unknown' 'Deceased' 'Republic' 'Jedi']
 ['Yoda' 'Unknown' '~880' 'Republic' 'Jedi']]

First 5 character names:
['Anakin Skywalker' 'Padmé Amidala' 'Obi-Wan Kenobi' 'Qui-Gon Jinn' 'Yoda']

First 10 character names and ages:
[['Anakin Skywalker' '22']
 ['Padmé Amidala' '27']
 ['Obi-Wan Kenobi' '~38']
 ['Qui-Gon Jinn' 'Deceased']
 ['Yoda' '~880']
 ['Mace Windu' '~53']
 ['Palpatine / Darth Sidious' '~65']
 ['Count Dooku / Darth Tyranus' '~83']
 ['Countess Dooku (cameo)' '~80+']
 ['Bail Organa' '~45']]


In [14]:
# Extract Age column

age_text = data[:, 2]

print("Age column:")
print(age_text)

Age column:
['22' '27' '~38' 'Deceased' '~880' '~53' '~65' '~83' '~80+' '~45' '~30'
 '~60' 'Deceased' '~50' '~30' '~40' '~30–40' '~30' '~60' '0 (as droid)'
 'Unknown' '~40' '13' '~40–50' '~60' '~60' '~55' '~50' '~50' 'Unknown'
 '~60' '~50' '~35' '~40' '~30' '~50' '~20' '~50' '~50' '~60' 'Deceased'
 '~10 (20 biological)' '~10' '~150' '~200' '~60' '~30–35' '~50' 'Unknown'
 'Unknown' '~60' '~30' '~60' '~60' '~80' '~40' '~60' '~50' '~10' '~10'
 '~10' '~10' '~10' '~30' '~30–40' '~40–50' '~300' '~500' '~40' '~20–30'
 '~15' '~60' '~55' '~25–30' '~60' '~20' '~20' '~20' '~?' '~50' '~?'
 '~20–30' '~50' '~200' '~40' '~40' '~40' '~80' '~40–50' '~?' '~30'
 '~30–40' '~50' '~45–50']


In [15]:
# Clean the Age column

age_values = []

for age in age_text:
    age = age.strip()

    # Remove "~"
    age = age.replace("~", "")

    # Handle ranges such as 30–40
    if "–" in age:
        parts = age.split("–")
        try:
            first = float(parts[0])
            second = float(parts[1])
            age_values.append((first + second) / 2)
        except:
            age_values.append(np.nan)

    # Handle values such as "10 (20 biological)"
    elif "(" in age:
        try:
            age_values.append(float(age.split("(")[0].strip()))
        except:
            age_values.append(np.nan)

    # Handle "80+"
    elif "+" in age:
        try:
            age_values.append(float(age.replace("+", "")))
        except:
            age_values.append(np.nan)

    # Handle normal numbers
    else:
        try:
            age_values.append(float(age))
        except:
            age_values.append(np.nan)

age_values = np.array(age_values)

print("Cleaned age values:")
print(age_values)

Cleaned age values:
[ 22.   27.   38.    nan 880.   53.   65.   83.   80.   45.   30.   60.
   nan  50.   30.   40.   35.   30.   60.    0.    nan  40.   13.   45.
  60.   60.   55.   50.   50.    nan  60.   50.   35.   40.   30.   50.
  20.   50.   50.   60.    nan  10.   10.  150.  200.   60.   32.5  50.
   nan   nan  60.   30.   60.   60.   80.   40.   60.   50.   10.   10.
  10.   10.   10.   30.   35.   45.  300.  500.   40.   25.   15.   60.
  55.   27.5  60.   20.   20.   20.    nan  50.    nan  25.   50.  200.
  40.   40.   40.   80.   45.    nan  30.   35.   50.   47.5]


In [16]:
# Mathematical operations

valid_ages = age_values[~np.isnan(age_values)]

print("Sum of ages:", np.sum(valid_ages))
print("Minimum age:", np.min(valid_ages))
print("Maximum age:", np.max(valid_ages))
print("Average age:", np.mean(valid_ages))

Sum of ages: 5433.5
Minimum age: 0.0
Maximum age: 880.0
Average age: 64.68452380952381


In [17]:
# Element-wise mathematical operations

print("First 10 ages:", valid_ages[:10])

print("\nAges + 5:")
print(valid_ages[:10] + 5)

print("\nAges × 2:")
print(valid_ages[:10] * 2)

print("\nAges squared:")
print(valid_ages[:10] ** 2)

First 10 ages: [ 22.  27.  38. 880.  53.  65.  83.  80.  45.  30.]

Ages + 5:
[ 27.  32.  43. 885.  58.  70.  88.  85.  50.  35.]

Ages × 2:
[  44.   54.   76. 1760.  106.  130.  166.  160.   90.   60.]

Ages squared:
[4.840e+02 7.290e+02 1.444e+03 7.744e+05 2.809e+03 4.225e+03 6.889e+03
 6.400e+03 2.025e+03 9.000e+02]


In [18]:
# Create a 2D age array for axis-wise operations

age_matrix = valid_ages[:60].reshape(3, 20)

print("Age matrix:")
print(age_matrix)

print("\nSum across rows (axis=1):")
print(np.sum(age_matrix, axis=1))

print("\nSum across columns (axis=0):")
print(np.sum(age_matrix, axis=0))

Age matrix:
[[ 22.   27.   38.  880.   53.   65.   83.   80.   45.   30.   60.   50.
   30.   40.   35.   30.   60.    0.   40.   13. ]
 [ 45.   60.   60.   55.   50.   50.   60.   50.   35.   40.   30.   50.
   20.   50.   50.   60.   10.   10.  150.  200. ]
 [ 60.   32.5  50.   60.   30.   60.   60.   80.   40.   60.   50.   10.
   10.   10.   10.   10.   30.   35.   45.  300. ]]

Sum across rows (axis=1):
[1681.  1135.  1042.5]

Sum across columns (axis=0):
[127.  119.5 148.  995.  133.  175.  203.  210.  120.  130.  140.  110.
  60.  100.   95.  100.  100.   45.  235.  513. ]


In [19]:
# Statistical operations

print("Mean age:", np.mean(valid_ages))
print("Median age:", np.median(valid_ages))
print("Standard deviation:", np.std(valid_ages))
print("Variance:", np.var(valid_ages))

Mean age: 64.68452380952381
Median age: 45.0
Standard deviation: 110.42016022582584
Variance: 12192.61178429705


In [20]:
# Reshaping

print("Original shape:", age_matrix.shape)

reshaped_age_matrix = age_matrix.reshape(4, 15)

print("New shape:", reshaped_age_matrix.shape)
print("\nReshaped age matrix:")
print(reshaped_age_matrix)

Original shape: (3, 20)
New shape: (4, 15)

Reshaped age matrix:
[[ 22.   27.   38.  880.   53.   65.   83.   80.   45.   30.   60.   50.
   30.   40.   35. ]
 [ 30.   60.    0.   40.   13.   45.   60.   60.   55.   50.   50.   60.
   50.   35.   40. ]
 [ 30.   50.   20.   50.   50.   60.   10.   10.  150.  200.   60.   32.5
   50.   60.   30. ]
 [ 60.   60.   80.   40.   60.   50.   10.   10.   10.   10.   10.   30.
   35.   45.  300. ]]


In [21]:
# Broadcasting

print("Original first row:")
print(reshaped_age_matrix[0])

print("\nAfter adding 5 to every age:")
print(reshaped_age_matrix[0] + 5)

print("\nAfter multiplying every age by 2:")
print(reshaped_age_matrix[0] * 2)

Original first row:
[ 22.  27.  38. 880.  53.  65.  83.  80.  45.  30.  60.  50.  30.  40.
  35.]

After adding 5 to every age:
[ 27.  32.  43. 885.  58.  70.  88.  85.  50.  35.  65.  55.  35.  45.
  40.]

After multiplying every age by 2:
[  44.   54.   76. 1760.  106.  130.  166.  160.   90.   60.  120.  100.
   60.   80.   70.]


In [22]:
# Save NumPy array

np.save("star_wars_ages.npy", valid_ages)

print("NumPy array saved successfully.")

NumPy array saved successfully.


In [23]:
# Load the saved NumPy array

loaded_ages = np.load("star_wars_ages.npy")

print("Loaded NumPy array:")
print(loaded_ages)

print("\nArray shape:", loaded_ages.shape)

Loaded NumPy array:
[ 22.   27.   38.  880.   53.   65.   83.   80.   45.   30.   60.   50.
  30.   40.   35.   30.   60.    0.   40.   13.   45.   60.   60.   55.
  50.   50.   60.   50.   35.   40.   30.   50.   20.   50.   50.   60.
  10.   10.  150.  200.   60.   32.5  50.   60.   30.   60.   60.   80.
  40.   60.   50.   10.   10.   10.   10.   10.   30.   35.   45.  300.
 500.   40.   25.   15.   60.   55.   27.5  60.   20.   20.   20.   50.
  25.   50.  200.   40.   40.   40.   80.   45.   30.   35.   50.   47.5]

Array shape: (84,)


In [24]:
# Performance comparison: Python list vs NumPy array

import time

# Create larger datasets for a meaningful comparison
python_list = valid_ages.tolist() * 10000
numpy_array = np.tile(valid_ages, 10000)

# Python list operation
start = time.perf_counter()

python_result = [x * 2 for x in python_list]

python_time = time.perf_counter() - start

# NumPy array operation
start = time.perf_counter()

numpy_result = numpy_array * 2

numpy_time = time.perf_counter() - start

print("Number of elements:", len(python_list))
print("Python list time:", python_time, "seconds")
print("NumPy array time:", numpy_time, "seconds")

print("NumPy speedup:", python_time / numpy_time, "times")

Number of elements: 840000
Python list time: 0.059769765000055486 seconds
NumPy array time: 0.00426329199990505 seconds
NumPy speedup: 14.019627321184345 times


In [25]:
# Verify saved and loaded arrays

print("Original array shape:", valid_ages.shape)
print("Loaded array shape:", loaded_ages.shape)

print("Arrays are equal:", np.array_equal(valid_ages, loaded_ages))

Original array shape: (84,)
Loaded array shape: (84,)
Arrays are equal: True


In [26]:
# Project 1 Summary

print("===== NumPy Data Explorer =====")
print("Dataset: Star Wars Characters")
print("Total rows:", data.shape[0])
print("Total columns:", data.shape[1])
print("Valid age values:", len(valid_ages))
print("Average age:", np.mean(valid_ages))
print("Median age:", np.median(valid_ages))
print("Minimum age:", np.min(valid_ages))
print("Maximum age:", np.max(valid_ages))
print("Saved array shape:", loaded_ages.shape)
print("Arrays saved and loaded successfully:", np.array_equal(valid_ages, loaded_ages))
print("NumPy performance speedup:", round(python_time / numpy_time, 2), "times")

===== NumPy Data Explorer =====
Dataset: Star Wars Characters
Total rows: 94
Total columns: 5
Valid age values: 84
Average age: 64.68452380952381
Median age: 45.0
Minimum age: 0.0
Maximum age: 880.0
Saved array shape: (84,)
Arrays saved and loaded successfully: True
NumPy performance speedup: 14.02 times


# Syntecxhub - NumPy Data Explorer

## Project 1

### Dataset
Star Wars Characters

### Topics Covered
- NumPy array creation and loading
- Indexing and slicing
- Mathematical operations
- Axis-wise operations
- Statistical operations
- Reshaping
- Broadcasting
- Saving and loading NumPy arrays
- Python list vs NumPy performance comparison

### Final Results
- Dataset rows: 94
- Dataset columns: 5
- Valid age values: 84
- Average age: 64.68
- Median age: 45
- Minimum age: 0
- Maximum age: 880
- NumPy performance speedup: 14.02×